File: BubbleSort/BubbleSort.py
#https://www.pythoncentral.io/pythons-range-function-explained/
def bubbleSort(aList):
    #parameters for range func to generate broader loop index i
    start = len(aList) - 1
    #use len(aList)-1 instead of start = len(aList),
    #otherwise aList[j+1] would cause 'list index out of range'
    stop = 0
    step = -1 #each loop decrease by 1
   
    #broader loop
    print "array length: ", len(aList)
    for i in range(start, stop, step):
        #inner loop
        print "loop index i : ", i
        for j in range(i):
            print "innder index j: ", j
            if aList[j] > aList[j+1]:
                #swap
                print "before swap: aList[j]: aList[j+1]:", aList[j], aList[j+1]
                tmp = aList[j]
                aList[j] = aList[j+1]
                aList[j+1] = tmp
                print "after swap: aList[j]: aList[j+1]:", aList[j], tmp
        
def bubbleSortSolution(alist):
    for passnum in range(len(alist)-1,0,-1):
        for i in range(passnum):
            if alist[i]>alist[i+1]:
                #swap
                temp = alist[i]
                alist[i] = alist[i+1]
                alist[i+1] = temp

def bubbleSortTest(aList):
  start = len(aList) - 1
  stop = 0
  step = -1
  #outer loop, index i
  for i in range(start, stop, step):
    #inner loop, index j
    for j in range(i):
      if aList[j] > aList[j+1]:
        #swap
        tmp = aList[j]
        aList[j] = aList[j+1]
        aList[j+1] = tmp


def bulleSortUpdate(aList):
    length = len(aList)
    
    if length < 2:
        return aList
    
    start = length-1#mistake made, should be length-1 instead of length
    stop = 0
    step = -1

    #outter loop, i would be ... 3, 2, 1
    for i in range(start, stop, step):
        #inner loop from 0 to the value of i
        for j in range(i):
            if aList[j] > aList[j+1]:
                #swap
                tmp = aList[j]
                aList[j] = aList[j+1]
                aList[j+1] = tmp
    return aList
aList = [54,-17,26,93,-17,77,0,0,44,55,20]

updated = bulleSortUpdate(aList)

print updated

File: BubbleSort/BubbleSort.go
package main

import (
	"fmt"
)

func bubbleSortTest(aList []int) {
	for i := 1; i < len(aList); i++ {
		for j := 0; j < i; j++ {
			if aList[j] > aList[j+1] {
				//swap
				aList[j], aList[j+1] = aList[j+1], aList[j]
			}
		}
	}
}

func main() {
	aList := []int{103, -54, 26, 93, 0, -17, 77, 1, 17, 31, 44, 55, 20}
	bubbleSortUpdate(aList)
	fmt.Println(aList)
}

func bubbleSort(aSlice []int) {
	for i := len(aSlice) - 1; i > 0; i-- {
		for j := 0; j < i; j++ {
			if aSlice[j] > aSlice[j+1] {
				//swap
				aSlice[j], aSlice[j+1] = aSlice[j+1], aSlice[j]
			}
		}
	}
}

func bubbleSortUpdate(aList []int) []int {
	length := len(aList)
	if length < 2 {
		return aList
	}
	//outter loop i = length-1, ..., 2, 1
	for i := length - 1; i > 0; i-- {
		//inner loop j = 0, 1, ... i
		for j := 0; j < i; j++ {
			if aList[j] > aList[j+1] {
				//swap
				aList[j], aList[j+1] = aList[j+1], aList[j]
			}
		}
	}
	return aList
}
